// giget/dist/shared/giget.3522139c.cjs
'use strict';

const promises = require('node:fs/promises');
const node_os = require('node:os');
const node_fs = require('node:fs');
const tar = require('tar');
const pathe = require('pathe');
const defu = require('defu');
const node_stream = require('node:stream');
const node_child_process = require('node:child_process');
const node_util = require('node:util');
const fetch = require('node-fetch-native');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e["default"] : e; }

const fetch__default = /*#__PURE__*/_interopDefaultLegacy(fetch);

async function download(url, filePath, opts = {}) {
  const infoPath = filePath + ".json";
  const info = JSON.parse(await promises.readFile(infoPath, "utf8").catch(() => "{}"));
  const headRes = await fetch.fetch(url, { method: "HEAD", headers: opts.headers }).catch(() => null);
  const etag = headRes?.headers.get("etag");
  if (info.etag === etag && node_fs.existsSync(filePath)) {
    return;
  }
  info.etag = etag;
  const res = await fetch.fetch(url, { headers: opts.headers });
  if (res.status >= 400) {
    throw new Error(`Failed to download ${url}: ${res.status} ${res.statusText}`);
  }
  const stream = node_fs.createWriteStream(filePath);
  await node_util.promisify(node_stream.pipeline)(res.body, stream);
  await promises.writeFile(infoPath, JSON.stringify(info), "utf8");
}
const inputRegex = /^(?<repo>[\w.-]+\/[\w.-]+)(?<subdir>[^#]+)?(?<ref>#[\w.-]+)?/;
function parseGitURI(input) {
  const m = input.match(inputRegex)?.groups;
  return {
    repo: m.repo,
    subdir: m.subdir || "/",
    ref: m.ref ? m.ref.substring(1) : "main"
  };
}
function debug(...args) {
  if (process.env.DEBUG) {
    console.debug("[giget]", ...args);
  }
}
function currentShell() {
  if (process.env.SHELL) {
    return process.env.SHELL;
  }
  if (process.platform === "win32") {
    return "cmd.exe";
  }
  return "/bin/bash";
}
function startShell(cwd) {
  cwd = pathe.resolve(cwd);
  const shell = currentShell();
  console.info(`(experimental) Opening shell in ${pathe.relative(process.cwd(), cwd)}...`);
  node_child_process.spawnSync(shell, [], {
    cwd,
    shell: true,
    stdio: "inherit"
  });
}

const github = (input, opts) => {
  const parsed = parseGitURI(input);
  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: { Authorization: opts.auth ? `Bearer ${opts.auth}` : void 0 },
    url: `https://github.com/${parsed.repo}/tree/${parsed.ref}${parsed.subdir}`,
    tar: `https://github.com/${parsed.repo}/archive/${parsed.ref}.tar.gz`
  };
};
const gitlab = (input, opts) => {
  const parsed = parseGitURI(input);
  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: { Authorization: opts.auth ? `Bearer ${opts.auth}` : void 0 },
    url: `https://gitlab.com/${parsed.repo}/tree/${parsed.ref}${parsed.subdir}`,
    tar: `https://gitlab.com/${parsed.repo}/-/archive/${parsed.ref}.tar.gz`
  };
};
const bitbucket = (input, opts) => {
  const parsed = parseGitURI(input);
  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: { Authorization: opts.auth ? `Bearer ${opts.auth}` : void 0 },
    url: `https://bitbucket.com/${parsed.repo}/src/${parsed.ref}${parsed.subdir}`,
    tar: `https://bitbucket.org/${parsed.repo}/get/${parsed.ref}.tar.gz`
  };
};
const sourcehut = (input, opts) => {
  const parsed = parseGitURI(input);
  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: { Authorization: opts.auth ? `Bearer ${opts.auth}` : void 0 },
    url: `https://git.sr.ht/~${parsed.repo}/tree/${parsed.ref}/item${parsed.subdir}`,
    tar: `https://git.sr.ht/~${parsed.repo}/archive/${parsed.ref}.tar.gz`
  };
};
const providers = {
  github,
  gh: github,
  gitlab,
  bitbucket,
  sourcehut
};

const DEFAULT_REGISTRY = "https://raw.githubusercontent.com/unjs/giget/main/templates";
const registryProvider = (registryEndpoint = DEFAULT_REGISTRY) => {
  return async (input) => {
    const start = Date.now();
    const registryURL = `${registryEndpoint}/${input}.json`;
    const res = await fetch__default(registryURL);
    if (res.status >= 400) {
      throw new Error(`Failed to download ${input} template info from ${registryURL}: ${res.status} ${res.statusText}`);
    }
    const info = await res.json();
    if (!info.tar || !info.name) {
      throw new Error(`Invalid template info from ${registryURL}. name or tar fields are missing!`);
    }
    debug(`Fetched ${input} template info from ${registryURL} in ${Date.now() - start}ms`);
    return info;
  };
};

const sourceProtoRe = /^([\w-.]+):/;
async function downloadTemplate(input, opts = {}) {
  opts = defu.defu({
    registry: process.env.GIGET_REGISTRY,
    auth: process.env.GIGET_AUTH
  }, opts);
  const registry = opts.registry !== false ? registryProvider(opts.registry) : null;
  let providerName = opts.provider || (registryProvider ? "registry" : "github");
  let source = input;
  const sourceProvierMatch = input.match(sourceProtoRe);
  if (sourceProvierMatch) {
    providerName = sourceProvierMatch[1];
    source = input.substring(sourceProvierMatch[0].length);
  }
  const provider = opts.providers?.[providerName] || providers[providerName] || registry;
  if (!provider) {
    throw new Error(`Unsupported provider: ${providerName}`);
  }
  const template = await Promise.resolve().then(() => provider(source, { auth: opts.auth })).catch((err) => {
    throw new Error(`Failed to download template from ${providerName}: ${err.message}`);
  });
  template.name = (template.name || "template").replace(/[^a-z0-9-]/gi, "-");
  template.defaultDir = (template.defaultDir || template.name).replace(/[^a-z0-9-]/gi, "-");
  const cwd = pathe.resolve(opts.cwd || ".");
  const extractPath = pathe.resolve(cwd, opts.dir || template.defaultDir);
  if (opts.forceClean) {
    await promises.rm(extractPath, { recursive: true, force: true });
  }
  if (!opts.force && node_fs.existsSync(extractPath) && node_fs.readdirSync(extractPath).length) {
    throw new Error(`Destination ${extractPath} already exists.`);
  }
  await promises.mkdir(extractPath, { recursive: true });
  const tmpDir = pathe.resolve(node_os.homedir(), ".giget", opts.provider, template.name);
  const tarPath = pathe.resolve(tmpDir, (template.version || template.name) + ".tar.gz");
  if (opts.preferOffline && node_fs.existsSync(tarPath)) {
    opts.offline = true;
  }
  if (!opts.offline) {
    await promises.mkdir(pathe.dirname(tarPath), { recursive: true });
    const s2 = Date.now();
    await download(template.tar, tarPath, { headers: template.headers }).catch((err) => {
      if (!node_fs.existsSync(tarPath)) {
        throw err;
      }
      debug("Download error. Using cached version:", err);
      opts.offline = true;
    });
    debug(`Downloaded ${template.tar} to ${tarPath} in ${Date.now() - s2}ms`);
  }
  if (!node_fs.existsSync(tarPath)) {
    throw new Error(`Tarball not found: ${tarPath} (offline: ${opts.offline})`);
  }
  const s = Date.now();
  const subdir = template.subdir?.replace(/^\//, "") || "";
  await tar.extract({
    file: tarPath,
    cwd: extractPath,
    onentry(entry) {
      entry.path = entry.path.split("/").splice(1).join("/");
      if (subdir) {
        if (entry.path.startsWith(subdir + "/")) {
          entry.path = entry.path.substring(subdir.length);
        } else {
          entry.path = "";
        }
      }
    }
  });
  debug(`Extracted to ${extractPath} in ${Date.now() - s}ms`);
  return {
    ...template,
    source,
    dir: extractPath
  };
}

exports.downloadTemplate = downloadTemplate;
exports.registryProvider = registryProvider;
exports.startShell = startShell;
